components/datetime/src/format/datetime.rs

```diff
@@ -221,23 +230,62 @@ where
             field.length,
         )?,
         FieldSymbol::DayPeriod(period) => {
-            let symbol = symbols.get_symbol_for_day_period(
-                period,
-                field.length,
-                datetime.datetime().hour().ok_or(Error::MissingInputField)?,
-                arithmetic::is_top_of_hour(
-                    pattern,
-                    datetime.datetime().minute().map(u8::from).unwrap_or(0),
-                    datetime.datetime().second().map(u8::from).unwrap_or(0),
-                ),
-            );
+            let symbol = symbols
+                .expect("Expect symbols to be present")
+                .get_symbol_for_day_period(
+                    period,
+                    field.length,
+                    datetime.datetime().hour().ok_or(Error::MissingInputField)?,
+                    arithmetic::is_top_of_hour(
+                        pattern,
+                        datetime.datetime().minute().map(u8::from).unwrap_or(0),
+                        datetime.datetime().second().map(u8::from).unwrap_or(0),
+                    ),
+                );
             w.write_str(symbol)?
         }
         field @ FieldSymbol::TimeZone(_) => return Err(Error::UnsupportedField(field)),
     };
     Ok(())
 }
 
+// This function determins whether the struct will load symbols data.
+// Keep it in sync with the `write_field` use of symbols.
+pub fn analyze_pattern(pattern: &Pattern, supports_time_zones: bool) -> Result<bool, &Field> {
+    let fields = pattern.items().iter().filter_map(|p| match p {
+        PatternItem::Field(field) => Some(field),
+        _ => None,
+    });
+
+    let mut requires_symbols = false;
+
+    for field in fields {
+        if !requires_symbols {
+            requires_symbols = match field.symbol {
+                FieldSymbol::Month(_) => {
+                    !matches!(field.length, FieldLength::One | FieldLength::TwoDigit)
+                }
+                FieldSymbol::Weekday(_) | FieldSymbol::DayPeriod(_) => true,
+                _ => false,
+            }
+        }
+
+        if supports_time_zones {
+            if requires_symbols {
+                // If we require time zones, and symbols, we know all
+                // we need to return already.
+                break;
+            }
+        } else if matches!(field.symbol, FieldSymbol::TimeZone(_)) {
+            // If we don't support time zones, and encountered a time zone
+            // field, error out.
+            return Err(field);
+        }
```

Comment:
Suggestion: I don't see the point of this if/elseif.  More specifically, I don't think checking `supports_time_zones` affects any behavior.  I think you can remove the `supports_time_zones` argument and just unconditionally break whenever `requires_symbols` gets set (or, better, delete that variable and just `return true`).

Reply:
The value is that we also validate the time zone availability here. If I were to apply the change you're suggesting, assuming I understand it, then we wouldn't error out if we established we need symbols but also the pattern used time zone and was called from non-time-zoned DTF.

Am I missing something?

Reply:
ok, I didn't catch that you actually wanted to validate something.  This is fine then.